lib/internal/process/warning.js

```diff
@@ -116,24 +116,23 @@ function setupProcessWarnings() {
       ctor = type;
       code = undefined;
       type = 'Warning';
+    } else if (type !== undefined && typeof type !== 'string') {
+      throw new ERR_INVALID_ARG_TYPE('type', 'string', type);
     }
     if (typeof code === 'function') {
       ctor = code;
       code = undefined;
-    }
-    if (code !== undefined && typeof code !== 'string')
+    } else if (code !== undefined && typeof code !== 'string') {
       throw new ERR_INVALID_ARG_TYPE('code', 'string', code);
-    if (type !== undefined && typeof type !== 'string')
-      throw new ERR_INVALID_ARG_TYPE('type', 'string', type);
-    if (warning === undefined || typeof warning === 'string') {
+    }
+    if (typeof warning === 'string') {
```

Comment:
This removes the possibility of an `undefined` warning that has the other properties specified.

Reply:
While not documented, as far as I can tell this was requested in https://github.com/nodejs/node/pull/4782

Reply:
`process.emitWarning()` produces: `(node:XXXX) Warning`. I am not convinced that this is useful at all. This seems like a bugfix as no args should result in an error. Therefore I would say it is a patch. However, I would also keep this if it makes this semver-major.

Ping @nodejs/tsc please alos give an opinion.

Reply:
> `process.emitWarning()` produces: `(node:XXXX) Warning`. I am not convinced that this is useful at all. 

`process.emitWarning(undefined, 'fhqwhgads', 'come-on')` produces the slightly more useful `(node:91812) [come-on] fhqwhgads`. 

I have no idea if anyone uses that pattern or not. Maybe it's time for me to figure out how to use Gzemnid.

In the spirit of the [onboarding-extras.md](https://github.com/nodejs/node/blob/master/doc/onboarding-extras.md):

> be conservative – that is, if a change has the remote chance of breaking something, go for semver-major

It seems like this should be semver-major as it has a "remote change of breaking something".

(If we're not going to call this semver-major, I won't push back, but we really ought to remove the quoted sentence above from the doc since it is clearly not something we abide by in that case.)